src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs

```diff
@@ -22,6 +23,9 @@ public static class ConcurrencyUtilities
                 throw new ArgumentNullException(nameof(filePath));
             }
 
+            // limit the number of unauthorized, this should be around 30 seconds.
+            var unauthorizedAttemptsLeft = 3000;
```

Comment:
Can't we retry for a specific duration (using a stopwatch) rather than on a count?